File: backend/app/schemas/dataset.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class DatasetProfile(BaseModel):
    missing_values: dict[str, int]  
    missing_percentage: dict[str, float]
    duplicate_rows: int
    column_types: dict[str, str]  


class DatasetResponse(BaseModel):
    dataset_uuid: str
    original_filename: str
    rows: int
    columns: int
    file_size: int
    uploaded_at: datetime

    column_names: list[str]
    numeric_columns: list[str]
    categorical_columns: list[str]
    profile: DatasetProfile

File: frontend/src/types/dataset.ts
export interface DatasetProfile {
    missing_values: Record<string, number>;
    missing_percentage: Record<string, number>;
    duplicate_rows: number;
    column_types: Record<string, number>;

    numeric_statistics: Record<string, Record<string, number | string | null >>;
    categorical_statistics: Record<string, Record<string, number | string | null>>;
}

export interface DatasetResponse {
    dataset_uuid: string;
    original_filename: string;
    rows: number;
    columns: number;
    file_size: number;
    uploaded_at: string;
    column_names: string[];
    numeric_columns: string[];
    categorical_columns: string[];
    preview: Record<string, unknown>[];
    profile: DatasetProfile;
}